roundabout quick-fix feature

kianzarrin opened the pull request:
Fixes #539. 
Phase 2:  #578 

EDIT: Tool: priority tool
CTRL-shift-click auto detects roundabout. If detected roundabout rules are applied. if not i fall back to the original basic priority mass edit code.

Path finding for roundabout follows  oneway roads and prefers forward then left then right. It avoids segments that join the middle of search path.

traffic rule fixes are controlled by options:
 - dedicated exit lanes
 - stay in lane in roundabout
 - stay in line near roundabout if the node is too close. see https://github.com/Strdate/AutomaticRoundaboutBuilder/issues/8
 - enter blocked junction in round about
 - enter blocked junction at entry.
 - no pedasterian crossing to the center of the roundabout
 - no pedasterian crossing around the roundabout. 
 - highway merging rules mess up the roundabout. so I activate junction lane switching. This is the case even if highway rules are not enabled. This way if someone decides to enable highway rules they do not need to setup all the roundabouts again. (should we make highway rules more clever to detect roundabouts? related #569)
 - add priority signs. 
 
I created Mass edit options and appended them to the policies tab. I put it in a separate file so its easy to make it a separate tab if desired. I encapsulated checkbox option in a class to avoid writing many many lines of repeated code. So the style is a bit different than usual. This encapsulation may even be use as a step toward #562.

I created a static class called shortcut to easily access commonly used stuff like different managers, netSerice, GetNode, GetSegment, GetSegmentEnd etc.

Can someone kindly add these translation keys and teach me how to download them into my branch?
[crowdin.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3914136/crowdin.txt)

modified service manger and netService to add some shared code.

aubergine10 commented:
Which tool needs to be active to apply the one-click roundabout?

aubergine10 commented:
> Path finding for roundabout follows oneway roads and prefers forward then left then right.

I assume this inverts based on which side traffic drives on?

So if traffic is driving on left, it would prefer clockwise. If traffic drives on right, it would prefer anti-clockwise.

aubergine10 commented on TLM/TLM/Manager/Impl/LaneArrowManager.cs:
```diff
@@ -221,6 +237,44 @@ public class LaneArrowManager
+        }
+
         public static class SeparateTurningLanes{
             /// <summary>
```
Just noticed; this comment is in wrong place and wrong format, should be above the method.

kianzarrin replied:
It IS above the method. But under the class declaration. 
GIT hub diff tool makes things confusing. I usually download the code and use a powerful diff tool to review changes :).

aubergine10 commented on TLM/TLM/Manager/Impl/LaneArrowManager.cs:
```diff
@@ -297,20 +364,13 @@ public static class SeparateTurningLanes{
-                int leftLanesCount = CountTargetLanesTowardDirection(segmentId, nodeId, ArrowDirection.Left);
-                int rightLanesCount = CountTargetLanesTowardDirection(segmentId, nodeId, ArrowDirection.Right);
-                int forwardLanesCount = CountTargetLanesTowardDirection(segmentId, nodeId, ArrowDirection.Forward);
+                int leftLanesCount = LaneArrowManager.CountTargetLanesTowardDirection(segmentId, nodeId, ArrowDirection.Left);
```
Aren't these 3 lines of code going to repeat the iteration in `CountTargetLanesTowardDirection()` three times?

Would it be worth having new `TallyTargetLanesPerDirection()` method that iterates once and returns how many lanes are in each direction via `out` params?

kianzarrin replied:
This is not part of the critical path of the code. And its not the bottleneck either. The time spent here pales in comparison to the time spent calling the managers to apply junction restrictions/priority rules/... which are surprisingly slow! Optimizing this will have no noticeable influence on the speed of the code. But if you feel strongly about this I can Optimize it.

aubergine10 replied:
If there are bigger optimisation gains elsewhere, then we should focus on those.

kianzarrin commented on TLM/TLM/Manager/Impl/ExtNodeManager.cs:
```diff
@@ -29,6 +29,32 @@ public class ExtNodeManager
+        /// </summary>
+        /// <param name="nodeId"></param>
+        /// <returns></returns>
+        public static bool IsHighwayJunction(ushort nodeId) {
```
`JunctionHasOnlyHighwayRoads` is too specific. I will change it to `DoesJunctionFollowHighwayRules`. for now I just assumed (after reading the code of course) this is when all roads connected are highway.

aubergine10 replied:
Off-topic: I suspect that users who build their roundabouts out of highway road probably end up with some very strange stuff happening if they've got highway rules activated in mod options.

kianzarrin commented on TLM/TLM/Manager/Impl/ExtSegmentEndManager.cs:
```diff
@@ -176,6 +176,20 @@ public class ExtSegmentEndManager
             return CalculateArrowDirection(sourceDir, targetDir);
         }
 
+        public ArrowDirection GetDirection(ushort segmentId0, ushort segmentId1, ushort nodeId = 0) {
```
The doc is in the interface. This method calls the private function bellow to find the direction. Also look at the already existing GetDirection method above.

kianzarrin commented on TLM/TLM/Manager/Impl/LaneArrowManager.cs:
```diff
@@ -45,6 +46,21 @@ public class LaneArrowManager
 
             return false;
         }
+        public bool AddLaneArrows(uint laneId,
```
None of the already existing methods have documentation. I will try to add doc for all of them. @aubergine10 can you please tell me what exactly `bool overrideHighwayArrows = false` does in `SetLaneArrows()`?

aubergine10 replied:
Regarding `overrideHighwayArrows`, I have no idea; I assume that's code from original TM:PE repo by Victor.

kianzarrin replied:
I have no idea either. I didn't understood it when I read the code. OK I will add documentation but I will not document this argument.

kianzarrin commented:
> So if traffic is driving on left, it would prefer clockwise. If traffic drives on right, it would prefer anti-clockwise.

This is handled `GetSortedSegments()` while it does look for LHT I did make mistake. I will fix.

kvakvs commented:
@kianzarrin 
Added your strings, i assume this is the right PR for them
Renamed string keys and removed periods, don't use periods on form labels

```
menu.csv:
"Tooltip.Keybinds:Add priority signs" => "ctrl-shift-click:Quick setup" 

options.csv
"Roundabout.Option:Allocate dedicated exit lanes" => "Allocate dedicated exit lanes"
"Roundabout.Tooltip:Allocate dedicated exit lanes": "One dedicated lane for each exit, the rest of lanes go straight"
"Roundabout.Option:Stay in lane inside roundabout" => "Stay in lane inside the roundabout"
"Roundabout.Option:Stay in lane outside roundabout" => "Stay in lane on the roads approaching the roundabout"
"Roundabout.Tooltip:Stay in lane outside roundabout"=> "Vehicles shall not switch lanes too close to the roundabout"
"Roundabout.Option:No crossing inside" => "Pedestrians shall not cross to the center of roundabout"
"Roundabout.Option:No crossing on incoming roads" => "Pedestrians shall not cross the roads approaching the roundabout"
"Roundabout.Option:Set priority signs" => "Add priority signs on the roundabout junctions"
```

kvakvs commented on TLM/TLM/packages.config:
```diff
@@ -1,10 +1,5 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Microsoft.CodeAnalysis.FxCopAnalyzers" version="2.9.7" targetFramework="net35" developmentDependency="true" />
```
This file should not be modified, you essentially removed analyzers.

kianzarrin replied:
You have commented on the wrong diff. I accidentally added unwanted code to packages.config and then removed it. So all in all it has not changed.

krzychu124 commented on TLM/TLM/UI/Helpers/ISerializableOptionBase.cs:
```diff
@@ -0,0 +1,9 @@
+    public interface ILegacySerializableOption
+    {
+        public void Load(byte data); // TODO keep this for backward compatibality.
+        public byte Save(); // TODO: delete this once xml serialization is ready
```
You can't use `public` here.

kianzarrin replied:
OK I will change it to internal then @krzychu124  is that OK?

krzychu124 replied:
I am not sure why your IDE doesn't throw an error there. The only valid identifier is `new` at that point or not at all.

kianzarrin replied:
@krzychu124 why do I need new at point of definition? Are you commenting on the wrong file?

krzychu124 replied:
> why do I need new at point of definition?

You don't need `new` at this point, either public. I said that `new` it's the only valid and allowed keyword of interface member declaration which you can put there if you need (example below).

```cs
public interface ITest: IOther {
  new void DoSomething(); //you have to use 'new' here because it hides IOther.DoSomething declaration
}
public interface IOther {
  void DoSomething(); //you shouldn't add any modifier here, because no access modifiers are allowed on interface member declarations.
}
```

krzychu124 commented on TLM/TLM/State/OptionsTabs/OptionsMassEditTab.cs:
```diff
@@ -0,0 +1,58 @@
+
+        internal static void MakePanel_MassEdit(UIHelperBase panelHelper) {
+            UIHelperBase raboutGroup = panelHelper.AddGroup(
+                T("MassEdit.Group: Roundabouts"));
```
It should be `MassEdit.Group:Roundabouts`

kianzarrin commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -0,0 +1,312 @@
+            var list1 = GetSortedSegmentsHelper( headNodeId, segmentId, ArrowDirection.Left   ,  lht);
+            var list2 = GetSortedSegmentsHelper( headNodeId, segmentId, ArrowDirection.Right  , !lht);
+
+            list0.AddRange(list1);
```
Yes that is a bug but again you have commented on outdated code. The latest version has the fix and is tested.

krzychu124 commented on TLM/TLM/UI/Helpers/SerializableUIOptionBase.cs:
```diff
@@ -0,0 +1,42 @@
+        public virtual TVal Value {
+            get => (TVal)ValueField.GetValue(null);
+            set => ValueField.SetValue(null, value);
+        }
```
It doesn't look that bad. It's quite slow, but we don't set more than one checkbox value at a time. It should be ok 😉

krzychu124 commented:
Maybe it's stupid question, but i don't see any info in description. Should I expect some sort of visual feedback when I try to use `CTRL-shift-click` and roundabout was not detected? 
I'm asking because I think it's not working, even if I click on actual roundabout. 
I'll have to check settings, because maybe I changed something.

kianzarrin commented:
@krzychu124 
> Maybe it's stupid question, but i don't see any info in description. Should I expect some sort of visual feedback when I try to use `CTRL-shift-click` and roundabout was not detected?

For the time being is does nothing. The plan is to quick-fix priority road as part of  #541. You asked me to split the pull request. so I deleted the code for #541 . I commented out code bellow so that you know what is suppose to happen when I finish #541 .
```c#
            if (ctrlDown && shiftDown) {
                bool isRAbout = RoundaboutMassEdit.Instance.FixRabout(HoveredSegmentId);
                //if (!isRAbout) {
                //    // TODO uncomment as part of issue #541
                //    //PriorityRoad.FixRoad(HoveredSegmentId);
                //}
                if (isRAbout) {
                    RefreshMassEditOverlay();
                    return;
                }
            // TODO uncomment as part of issue #541
            //} else if (ctrlDown) {
            //
            //    //PriorityRoad.FixJunction(HoveredNodeId);
            //    RefreshMassEditOverlay();
            //    return;
            }
```

> I'm asking because I think it's not working, even if I click on actual roundabout.
> I'll have to check settings, because maybe I changed something.

You select the priority sign tool and ctrl+shift+click on a round about. It takes a while to setup all the rules (not my fault its the back-end interface that is slow) maybe a second. So if you want to see the changes keep holding ctrl for a while. as part of #541 I will make sure the mass-edit overlay remains active for a short while after releasing ctrl.

kvakvs commented:
When adding "stay in lane" to the upcoming lanes, don't do this in 1-lane road.
![bild](https://user-images.githubusercontent.com/288724/72663730-15480180-39f6-11ea-8365-6f1048363672.png)

kvakvs commented:
Play tested it a bit, i enjoy it.

The explanation of options in option tab is missing, entirely. We do not have UI guidelines on how to do it properly but a new person would never ever figure out where the options are and how to use them. Which means people will bombard forums and chat with questions, or entirely ignore most of these options.

kianzarrin commented:
> @kvakvs  The explanation of options in option tab is missing, entirely.
Are you talking about the tutorial?

kianzarrin commented:
> @kvakvs  When adding "stay in lane" to the upcoming lanes, don't do this in 1-lane road.
This is a problem in the lane connector. I created issue #617 for this. I think this is outside of the scope of this review.

kvakvs commented:
> > @kvakvs  The explanation of options in option tab is missing, entirely.
> > Are you talking about the tutorial?

The UX part, we do not have any explanations or docs how our options work. This is something to think about later, we want to explain every option. Tutorial popups don't do it good enough.

kvakvs commented on TLM/TLM/UI/SubTools/LaneConnectorTool.cs:
```diff
@@ -406,6 +371,49 @@ bool drawMarker
             }
         }
 
+        public static void StayInLane(ushort nodeId, StayInLaneMode stayInLaneMode = StayInLaneMode.Both) {
```
Extracted function: nice!

kvakvs commented on TLM/TLM/UI/SubTools/PrioritySignsTool.cs:
```diff
@@ -453,7 +517,7 @@ public PrioritySignsTool(TrafficManagerTool mainTool)
             Cleanup();
 
-            if (Options.prioritySignsOverlay) {
+            if (Options.prioritySignsOverlay ) {
```
A wild space appeared

aubergine10 commented:
Will be testing in-game in an hour or so, sorry for delay.

aubergine10 commented:
So my VS2019 is fubar, to the point where I've had to submit fault report to MS for assistance. I suggest just go ahead and merge this, I'll test when I've revived VS2019.
